### ISSUE TITLE
Fix increasing optimization slowness after system overhaul

### ISSUE THREAD

**Issue by wodeni** [penrose/penrose#116]
The new Style rewrite (#99) reduced the performance of the system. Based on the profiling results (#13), we suspect the optimizer becomes slower due to the increased number of transformations we do. We would like to explore other optimization methods that are of lower computational costs.

Some early ideas:
- eliminate the automatic differentiation
- use a zeroth order optimization method
  - it may converge slower overall but maybe each step will be faster
- numerically estimate gradients

**Comment by hypotext** [comment]
We might be able to get the optimization to converge much faster by sampling many initial states uniformly at random, then picking the lowest-energy one. This strategy will be a useful benchmark for smarter opt algorithms. See description of the random search algorithm [here](http://www.sft.asso.fr/Local/sft/dir/user-3775/documents/actes/Metti5_School/Lectures&Tutorials-Texts/Text-T2-Ruffio.pdf).

**Comment by wodeni** [comment]
Some comments:

- Note that `stack --profile` automatically includes __all top-level functions__ as cost centers. To undo that, you can use ` stack build --profile --ghc-options -fno-prof-auto`. See [GHC's official page on profiling](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/profiling.html) for more details. 
- I'm curious about the effect of `zeroGrads`. WIthout a good understanding of AD and these Autofloats I can't make any meaningful comment though. Intuitively I thought `r2f` worked in `zeroGrads` because the output `Obj`s have `Float` rather than `Autofloat a`?

**Comment by wodeni** [comment]
- ShapeDef
  - changed the shapes from record types to maps
  - looking up `shapedefs` (also, `shapedefs` seems to be recomputed every time)
  - the family of `get` functions are written inefficiently
    - test if records are really faster than maps: write a test script to load test retrieval of members
- Evaluator
  - Eval'ing the objective functions is slow.
    - looking obj/constr/comp redundantly (same in old system)
  - Redundant evaluation
- Finding gradient
  - Ask Edward Kmett about `reifyTape`, or performance of AD in general

**Comment by wodeni** [comment]
### How to set top-level `SCC` progma

- It requires higher version of ghc, which is 8.4.1 
- If we just run ghc on @wodeni's machine, there is an error coming from clang about the `-no-pie` flag. We fixed it by changing the setting of ghc in `/Library/Frameworks/GHC.framework/Versions/8.4.2-x86_64/usr/lib/ghc-8.4.2/settings`(solution found in: https://stackoverflow.com/questions/50386787/cabal-install-gcc-failed-in-phase-c-compiler/50419101#50419101). As a result, the following file compiles and generates correct .prof file:
```
main :: IO ()
main = do
    f

f  :: IO ()
f = do
    let xs = [1..1000000]
    let ys = [1..2000000]
    print $ {-# SCC last_xs #-} last xs
    print $ {-# SCC last_init_xs #-} last $ init xs
    print $ {-# SCC last_ys #-} last ys
    print $ {-# SCC last_init_ys #-} last $ init ys
{-# SCC f #-}
```
- To install 8.4.1 in stack, we needed to change the `resolver` in `stack.yaml`, which caused multiple rounds of errors. We eventually solved it by adding a flag: `allow-newer: true`. We also pasted in soe recommended `extra-deps` items for multiple times, but not sure which action fixed the error.

**Comment by wodeni** [comment]
### About the new system gradually slowing down

- We observed in the new system, without line search, that the stepping speed gradually slow down over time.
- To verify that this is a new phenomenon, we disabled the server and line search in a forked version of the old system (`HEAD` of `master`) and there seems to be no visually noticable difference of stepping speed over time.
- Therefore, the slow down over time situation happens exclusively in the new system. 
- __Why?__

**Comment by hypotext** [comment]
New info
* varyMap only needs to be constructed once per step but then the obj fn does not have the right type for autodiff
* constructing the varyMap inline causes a bit of slowdown (not that much) *and* doing all the varyMap lookups causes the rest of the slowdowns (checked by replacing the arg to evalFns M.empty)

Therefore, I believe the use of varyMap is the problem, not the construction. Possible solutions:

* can we mark GPIs as done to skip extraneous evaluation and lookups? (make sure they have the varying vals though...)
* maybe we should revert to the insertion but just written in a faster way. that seems less error-prone than the varyMap approach

**Comment by hypotext** [comment]
We started eliminating code from the system.
* genObjFn has empty args, no change
* changes happen when we remove the args of mkVaryMap. 
* hellos happen when mkVaryMap has type (Autofloat a) => [a] -> a
* hellos disappear when: 1) [a] is [String], 2) the return argument is a float, 3) [a] is an artificial single a, 4) mkVaryMap has no args but still returns a varyMap
* we eliminated Data.Map and we eliminated list
* it's not about [a] in mkVaryMap, because if mkVaryMap has type Autofloat a => a -> a, when we pass in the head of varyingVals (instead of the hardcoded float 0.0), we DO get hellos, whereas we did not with hardcoded floats

**Comment by hypotext** [comment]
Testing partial.hs (minimal version with record with polymorphism)

With wrong code:
1 step -> 1 hello (+2)
2 steps -> 3 hellos (+4)
3 steps -> 7 hellos (+8)
4 steps -> 15 hellos (+16)
5 steps -> 31 hellos

With right code:
number of hellos = number of steps

* If we make it a float, not poly., then we get the right number of hellos
* If we pull out the first step of the seq into a let binding, then we get the right number of hellos

**Comment by hypotext** [comment]
New TODO list

New list
- [ ] Write up stackoverflow question?
- [ ] Clean up new-style branches
- [ ] keep: stepsWithoutServer, varyMap, map keys -- keep them fast, make shapeDefs not use zipWithKeys
- [ ] clean up debug flags, redundant GPI evaluation
- [ ] consider monomorphizing the system to increase performance
- [ ] Look the nonconvergence issue
- [ ] rest of the port
- [ ] continuousMap, outsideOf
- [ ] Return Nimo's key

**Comment by hypotext** [comment]
Solved in `new-style-fast`. (But still not clear on _exactly_ why the two minimal examples with the "let" and without had such a dramatic speed/evaluation difference.)

**Comment by hypotext** [comment]
Asked a StackOverflow question [here](https://stackoverflow.com/questions/51829788/puzzling-performance-output-behavior-with-rank-2-polymorphism-in-haskell).

**Comment by hypotext** [comment]
I think I understand the StackOverflow answer better now. Just considering the minimal example, the pathological time/space behavior arose due to the interaction of many features across the compiler/language/PL spectrum:

* how GHC implements polymorphism and typeclasses
* GHC's typeclass defaulting behavior
* the implicit monomorphism GHC flag
* GHC compiler optimizations
* Haskell's lazy evaluation semantics (independent of compiler), combination with memoization,  interaction with side effects (printing "hello"), and interaction with strictness (`seq`)
* the common strategy of closure creation

Some notes on the minimal example:
- `forall a . (Fractional a) => a` is a function that takes **two** arguments, not one as I originally thought
- The execution becomes slower and slower, and the number of hellos doubles, because the slow version—for each call to `step`—needs to make _two_ more calls to `step` to evaluate `f Double (Fractional Double)`, because it's all happening under a lambda. What's happening here is very similar to the pathological [unmemoized polymorphic Fibonacci](https://stackoverflow.com/questions/11145503/why-does-adding-a-polymorphic-type-signature-degrade-performance) problem.
- Writing out some of the execution: (not totally sure about Haskell's execution semantics, so this may be wrong; also it's a "steps eventually," not a real small-step)
```
[env: s0 := State 100.0]
(vstate $ step $ step s0) Double (Fractional Double)

|-> expand the first step

[env: s0 := State 100.0]
(vstate $ [env: f1 := vstate $ step s0; vs1 := trace "h" f1] State (\t i -> (vs1 :: Double) `seq` f1 t i)) Double (Fractional Double)

|-> apply vstate to State

[env: s0 := State 100.0; f1 := vstate $ step s0; vs1 := trace "h" (f1 Double (Fractional Double))]
(\t i -> vs1 `seq` f1 t i) Double (Fractional Double)

|-> apply the function to its type and instance arguments

[env: s0 := State 100.0; f1 := vstate $ step s0; vs1 := trace "h" (f1 Double (Fractional Double))]
vs1 `seq` f1 Double (Fractional Double)

|-> evaluate vs1, memoizing along the way

[env: s0 := State 100.0; f1 := vstate $ step s0; vs1 := ?]
(trace "h" (f1 Double (Fractional Double)) `seq` f1 Double (Fractional Double))

|-> print "h"

[env: s0 := State 100.0; f1 := vstate $ step s0; vs1 := ?]
f1 Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> evaluate f1 (here's where memoization comes in, so we don't actually evaluate f1 twice!)

[env: s0 := State 100.0; f1 := ?; vs1 := ?]
(vstate $ step s0) Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> evaluate the left-hand side; inline step

[env: s0 := State 100.0; f1 := ?; vs1 := ?; vs0 = trace "h" s0]
(vstate $ State (vs0 `seq` s0)) Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> evaluate vstate $ State (...)

[env: s0 := State 100.0; f1 := ?; vs1 := ?; vs0 = trace "h" s0]
(vs0 `seq` s0) Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> evaluate vs0

[env: s0 := State 100.0; f1 := ?; vs1 := ?; vs0 = ?]
(trace "h" s0 `seq` s0) Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> print "h"

[env: s0 := State 100.0; f1 := ?; vs1 := ?; vs0 = ?]
(s0 `seq` s0) Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> evaluate s0, which is a value, so it's just looked up. and on the left side of the `seq`, the call defaults to Fractional Double, because the result isn't used. I'm not clear on how the `(State 100.0) Double (Fractional Double)` gets the type information inside the type wrapper

[env: s0 := State 100.0; f1 := ?; vs1 := ?; vs0 = ?]
((State 100.0) Double (Fractional Double) `seq` s0) Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> evaluate seq

[env: s0 := State 100.0; f1 := ?; vs1 := ?; vs0 = ?]
s0 Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> lookup s0

[env: s0 := State 100.0; f1 := ?; vs1 := ?; vs0 = ?]
(State 100.0) Double (Fractional Double) `seq` f1 Double (Fractional Double))

|-> same as before, evaluate seq, and now we know what vs0 *and* vs1 are (they stepped to values), and we know what `f1 Double (Fractional Double)` was (because that was vs1), so we can memoize all their values

[env: s0 := State 100.0; f1 := ?; f1 Double (Fractional Double) := (State 100.0) Double (Fractional Double); vs1 := (State 100.0) Double (Fractional Double); vs0 = (State 100.0) Double (Fractional Double)]
f1 Double (Fractional Double))

|-> now we simply lookup the value of `f1 Double (Fractional Double)` in the environment, since we have already evaluated f1 with those instances

(State 100.0) Double (Fractional Double)
```

Note, also, I'm not very familiar with writing lazy evaluation semantics that incorporate a closure/environment *and* forced strictness (`seq`). Does it evaluate outside-in or inside-out? Would be curious to see the correct formalism.

- When exactly are the closures called with type and typeclass instances? I'm not clear on this.

**Comment by hypotext** [comment]
Now that we've worked that out: how much of the pathological behavior exhibited by the minimal example actually applies to our system?

**I still don't know *exactly* what part of our system had exponential time complexity due to polymorphism**, since our `step` did have a let binding (unlike the bad example in the SO question). Could it be in `stepShapes` or deeper in the optimizer?

This is the [commit](https://github.com/penrose/penrose/commit/eaa08fde6f1f61a71eac24d840f7a94a1815273c) that resolved the slowness issue by making `varyingState :: [Float]`.

The SO answer suggested a very useful trick, which is to compile with `-Wtype-defaults`. I did so, and here are the results on two examples:

1. a [small example](https://gist.github.com/hypotext/786fa0fd46b9a2fa387e806f6bee3cbb) with autodiff, which seems to default to `Double`, but I had expected to see some of the autodiff floats in there.

2. most of [our system](https://gist.github.com/hypotext/4da36989a8d1ad8e2b5cd33d88f9bb39), which has a *lot* of defaulting behavior. I don't understand why it defaults to `‘Double’ Floating t0` and `‘Integer’ Num t0` in the line search.

We could still dig much deeper, for example to understand each example of defaulting behavior in our system, and to manually specialize each instance to ensure maximum speed / minimum unnecessary polymorphism (rather than using `r2f` on varyingState to make it a `Float`). I wonder if there is more slowness that could be easily removed if we tried this.
